src/include/network/connection_handle.h

```diff
@@ -0,0 +1,212 @@
+    using transition_result = std::pair<ConnState, action>;
+    /**
+     * Runs the internal state machine, starting from the symbol given, until no more
+     * symbols are available.
+     *
+     * Each state of the state machine defines a map from a transition symbol to an action
+     * and the next state it should go to. The actions can either generate the next symbol,
+     * which means the state machine will continue to run on the generated symbol, or signal
+     * that there is no more symbols that can be generated, at which point the state machine
+     * will stop running and return, waiting for an external event (user interaction, or system event)
+     * to generate the next symbol.
+     *
+     * @param action starting symbol
+     * @param connection the network connection object to apply actions to
+     */
+    void Accept(Transition action, ConnectionHandle &connection);
+
+  private:
+    /**
+     * delta is the transition function that defines, for each state, its behavior and the
+     * next state it should go to.
+     */
+    static transition_result Delta_(ConnState state, Transition transition);
+    ConnState current_state_ = ConnState::READ;
+  };
+
+  friend class StateMachine;
+  friend class ConnectionHandleFactory;
+
+  ConnectionHandle(int sock_fd, ConnectionHandlerTask *handler,
+                    std::shared_ptr<Buffer> rbuf, std::shared_ptr<Buffer> wbuf);
+
+  ProcessResult ProcessInitial();
+
+  /**
+   * Extracts the header of a Postgres start up packet from the read socket buffer
+   */
+  static bool ReadStartupPacketHeader(Buffer &rbuf, InputPacket &rpkt);
+
+  /**
+   * Routine to deal with the first packet from the client
+   */
+  void ProcessInitialPacket(InputPacket *pkt);
+
+  /**
+   * Routine to deal with general Startup message
+   */
+  void ProcessStartupPacket(InputPacket *pkt, int32_t proto_version);
+
+  /**
+   * Routine to deal with SSL request message
+   */
+  void ProcessSSLRequestPacket(InputPacket *pkt);
+
+  /**
+   * Writes a packet's header (type, size) into the write buffer
+   */
+  WriteState BufferWriteBytesHeader(OutputPacket *pkt);
+
+  /**
+   * Writes a packet's content into the write buffer
+   */
+  WriteState BufferWriteBytesContent(OutputPacket *pkt);
+
+  /**
+   * Used to invoke a write into the Socket, returns false if the socket is not
+   * ready for write
+   */
+  WriteState FlushWriteBuffer();
+
+  /**
+   * Set the socket to non-blocking mode
+   */
+  inline void SetNonBlocking(evutil_socket_t fd) {
+    auto flags = fcntl(fd, F_GETFL);
+    flags |= O_NONBLOCK;
+    if (fcntl(fd, F_SETFL, flags) < 0) {
+      LOG_ERROR("Failed to set non-blocking socket");
+    }
+  }
+
+  /**
+   * Set TCP No Delay for lower latency
+   */
+  inline void SetTCPNoDelay(evutil_socket_t fd) {
+    int one = 1;
+    setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof one);
+  }
+
+  int sock_fd_;                    // socket file descriptor
```

Comment:
Shouldn't we be using doxygen for class members too?

Reply:
Legacy. I can convert them but I don't see much pointer in generating docs for private things.